staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher_test.go

```diff
@@ -259,6 +305,150 @@ func TestWatchChanSync(t *testing.T) {
 	}
 }
 
+func TestWatchInitialEventsEndBookmark(t *testing.T) {
+	makePodFn := func(pref string) *example.Pod {
+		return &example.Pod{ObjectMeta: metav1.ObjectMeta{Name: fmt.Sprintf("foo-%s", pref), Namespace: "foo-ns"}}
+	}
+	makeBookmarkPodFn := func(rv string) *example.Pod {
+		return &example.Pod{ObjectMeta: metav1.ObjectMeta{ResourceVersion: rv, Annotations: map[string]string{"k8s.io/initial-events-end": "true"}}}
+	}
+	makeWatchEventFn := func(eventType watch.EventType, obj runtime.Object) watch.Event {
+		return watch.Event{Type: eventType, Object: obj}
+	}
+	createPodFn := func(ctx context.Context, t *testing.T, p *example.Pod, s storage.Interface) *example.Pod {
+		out := &example.Pod{}
+		objKey := fmt.Sprintf("/pods/%s/%s", p.Namespace, p.Name)
+		if err := s.Create(ctx, objKey, p, out, 0); err != nil {
+			t.Fatalf("Create failed: %v", err)
+		}
+		return out
+	}
+
+	scenarios := []struct {
```

Comment:
Overall, I'm not a fan of it.

With this PR, we are effectively unifying the semantics between etcd and watchcache.
So instead of adding this test, I would really like to see combining this test with the one you created for watchcache (https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go#L1614)
and moving it to pkg/storage/testing

and running the exact same test for the two implementations

Reply:
ah, okay, I totally forgot that we have tests that span multiple implementations. That make sense. Let me rework the test.